src/I18n/FormatterLocator.php

```diff
@@ -0,0 +1,91 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @copyright     Copyright (c) 2017 Aura for PHP
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.2.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\I18n;
+
+use Cake\I18n\Exception\I18nException;
+
+/**
+ * A ServiceLocator implementation for loading and retaining formatter objects.
+ *
+ * @internal
+ */
+class FormatterLocator
+{
+    /**
+     * A registry to retain formatter objects.
+     *
+     * @var array
+     */
+    protected $registry = [];
+
+    /**
+     * Tracks whether or not a registry entry has been converted from a
+     * callable to a formatter object.
+     *
+     * @var array
+     */
+    protected $converted = [];
+
+    /**
+     * Constructor.
+     *
+     * @param array $registry An array of key-value pairs where the key is the
+     * formatter name the value is a callable that returns a formatter object.
+     */
+    public function __construct(array $registry = [])
+    {
+        foreach ($registry as $name => $spec) {
+            $this->set($name, $spec);
+        }
+    }
+
+    /**
+     * Sets a formatter into the registry by name.
+     *
+     * @param string $name The formatter name.
+     * @param callable $spec A callable that returns a formatter object.
+     * @return void
+     */
+    public function set(string $name, callable $spec): void
```

Comment:
Are you still pushing to change callables closures?

Reply:
I am but method will practically never be used. The 2 formatters we have are already setup through the locator's constructor and I don't really see anyone needed to a new formatter type.

That said even using callables which return formatter instance is not necessary. We can just have a formatter name to class name making and the locator and create and return an instance.